File: Div3-Group1/PalindromeLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        bool answer= true;
        vector<int> container;

        ListNode* holder = head;

        if(holder->next==NULL)
        {
            return true;
        }
        else
        {
            while(holder)
            {
                container.push_back(holder->val);
                holder=holder->next;
            }
        }
        
        long long int start=0;
        long long int end=container.size()-1;

        while(start<end)
        {
            if(container[start]!=container[end])
            {
                answer=false;
                break;
            }   
            else{
                start++;
                end--;
            }
        }
        return answer;
    }
};

File: leet-code-solutions/palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head) -> Optional[ListNode]:
        # print(head,"head")
        prev = None
        cur = head
        while cur:
            cur.next, prev, cur = prev, cur, cur.next
        
        return prev
     


    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        temp=head
        counter=0

        while temp:
            temp=temp.next
            counter+=1
        
        if counter == 1:
            return True

        counter=(counter+1)//2

        after_mid_ptr=head
        before_mid_ptr=head

        for i in range(counter):
            after_mid_ptr=after_mid_ptr.next

        # reverse
        after_mid_ptr=self.reverseList(after_mid_ptr)

        # print(before_mid_ptr,"//",after_mid_ptr,counter)

        # for i in range(counter):
        #     # print(before_mid_ptr.val,after_mid_ptr.val)
        #     if before_mid_ptr.val!=after_mid_ptr.val:
        #         return False
        #     after_mid_ptr=after_mid_ptr.next
        #     before_mid_ptr=before_mid_ptr.next

        while before_mid_ptr and after_mid_ptr:
            if before_mid_ptr.val!=after_mid_ptr.val:
                return False
            after_mid_ptr=after_mid_ptr.next
            before_mid_ptr=before_mid_ptr.next
            
        return True